src/masternode-budget.cpp

```diff
@@ -450,14 +450,24 @@ std::string CBudgetManager::GetFinalizedBudgetStatus(const uint256& nHash) const
 
 bool CBudgetManager::AddFinalizedBudget(CFinalizedBudget& finalizedBudget)
 {
+    AssertLockNotHeld(cs_budgets);    // need to lock cs_main here (CheckCollateral)
     const uint256& nHash = finalizedBudget.GetHash();
+
     if (WITH_LOCK(cs_budgets, return mapFinalizedBudgets.count(nHash))) {
         LogPrint(BCLog::MNBUDGET,"%s: finalized budget %s already added\n", __func__, nHash.ToString());
         return false;
     }
 
-    if (!finalizedBudget.UpdateValid(GetBestHeight())) {
-        LogPrint(BCLog::MNBUDGET,"%s: Invalid finalized budget %s %s\n", __func__, nHash.ToString(), finalizedBudget.IsInvalidLogStr());
+    if (!finalizedBudget.IsWellFormed(GetTotalBudget(GetBestHeight()))) {
+        LogPrint(BCLog::MNBUDGET,"%s: invalid finalized budget: %s %s\n", __func__, nHash.ToString(), finalizedBudget.IsInvalidLogStr());
+        return false;
+    }
+
+    int nConf = 0;
+    std::string strError;
+    if (!CheckCollateral(finalizedBudget.GetFeeTXHash(), nHash, strError, finalizedBudget.nTime, nConf, true)) {
```

Comment:
Inside `UpdateValid` we are checking the finalized budget expiration and the collateral min confirmations. With this change, we are not checking them anymore here. What is the rationality behind?

Reply:
Here we are just decoupling the checks that need the current tip height from those that don't, and doing the static checks only in `AddFinalizedBudget`.
`UpdateValid` is still  in `SubmitFinalBudget` and `CheckAndRemove`. It will be moved back into  `UpdateValid` few commits later, once we cache the collateral tx height, and can perform the validation without locks.